component/remote/vault/refresher.go

```diff
@@ -0,0 +1,278 @@
+		return err
+	}
+
+	tm.readCounter.Inc()
+
+	tm.token = token
+
+	select {
+	case tm.onStateChange <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+// Run runs the tokenManager, blocking until the provided context is canceled.
+func (tm *tokenManager) Run(ctx context.Context) {
+	var cancelLifecycleWatcher context.CancelFunc
+	defer func() {
+		if cancelLifecycleWatcher != nil {
+			cancelLifecycleWatcher()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-tm.refreshTicker.Chan():
+			level.Info(tm.log).Log("msg", "refreshing token")
+			tm.updateToken(ctx)
+
+		case <-tm.onStateChange:
+			if cancelLifecycleWatcher != nil {
+				cancelLifecycleWatcher()
+			}
+
+			ctx, cancel := context.WithCancel(ctx)
+			cancelLifecycleWatcher = cancel
+
+			tm.updateLifecycleWatcher(ctx)
+		}
+	}
+}
+
+func (tm *tokenManager) updateToken(ctx context.Context) {
+	err := tm.updateTokenError(ctx)
+
+	if err != nil {
+		tm.updateHealth(component.Health{
+			Health:     component.HealthTypeUnhealthy,
+			Message:    fmt.Sprintf("failed to retrieve token: %s", err),
+			UpdateTime: time.Now(),
+		})
+	} else {
+		tm.updateHealth(component.Health{
+			Health:     component.HealthTypeUnhealthy,
+			Message:    "retrieved token",
+			UpdateTime: time.Now(),
+		})
+	}
+
+	tm.updateDebugInfo(time.Now())
+}
+
+func (tm *tokenManager) updateHealth(h component.Health) {
+	tm.healthMut.Lock()
+	defer tm.healthMut.Unlock()
+
+	tm.health = h
+}
+
+func (tm *tokenManager) updateDebugInfo(updateTime time.Time) {
+	tm.mut.RLock()
+	token := tm.token
+	tm.mut.RUnlock()
+
+	tm.debugMut.Lock()
+	defer tm.debugMut.Unlock()
+
+	tm.debugInfo = getSecretInfo(token, updateTime)
+}
+
+func (tm *tokenManager) updateLifecycleWatcher(ctx context.Context) {
+	tm.mut.RLock()
+	defer tm.mut.RUnlock()
+
+	if !needsLifecycleWatcher(tm.token) {
+		return
+	}
+
+	lw, err := tm.cli.NewLifetimeWatcher(&vault.LifetimeWatcherInput{
+		Secret:        tm.token,
+		RenewBehavior: vault.RenewBehaviorIgnoreErrors,
+	})
+	if err != nil {
+		level.Error(tm.log).Log("msg", "failed to create lifetime watcher, lease will not renew automatically", "err", err)
+		return
+	}
+
+	go lw.Start()
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				lw.Stop()
+				return
+
+			case <-lw.DoneCh():
+				if ctx.Err() != nil {
+					return
+				}
+				tm.updateToken(ctx)
+
+			case output := <-lw.RenewCh():
+				tm.refreshCounter.Inc()
+				level.Debug(tm.log).Log("msg", "token has renewed")
+				tm.updateDebugInfo(output.RenewedAt)
+			}
+		}
+	}()
+}
+
+// needsLifecycleWatcher determines if a secret needs a lifecycle watcher.
+// Secrets only need a lifecycle watcher if they are renewable or have a lease
+// duration.
+func needsLifecycleWatcher(secret *vault.Secret) bool {
+	if secret.Auth != nil {
+		return secret.Auth.Renewable || secret.Auth.LeaseDuration > 0
+	}
+	return secret.Renewable || secret.LeaseDuration > 0
+}
```

Comment:
This can also panic in case this is hit on the codepath of authManager using certain AuthMethods